### ISSUE TITLE
New contributions for KDE/Plasma software?

### ISSUE THREAD

**Issue by ratijas** [zsh-users/zsh-completions#984]
Hi,

I'm an active KDE developer, and I recently [wrote](https://invent.kde.org/plasma/plasma-sdk/-/merge_requests/56) [a lot](https://invent.kde.org/sdk/kdesrc-build/-/merge_requests/165) [of](https://invent.kde.org/system/dolphin/-/merge_requests/473) [zsh](https://invent.kde.org/plasma/plasma-workspace/-/merge_requests/2442) [completions](https://invent.kde.org/plasma/systemsettings/-/merge_requests/175) [for the](https://github.com/KDAB/GammaRay/pull/753) [commands](https://invent.kde.org/utilities/konsole/-/merge_requests/791) [I use](https://invent.kde.org/plasma/libkscreen/-/merge_requests/79) often, and I hope more are to come. Those completions are as intelligent and integrated as my level of understanding compdef features allowed me to write.

However, at some point we had a [disagreement](https://invent.kde.org/graphics/spectacle/-/merge_requests/174#note_584783) between our [visions](https://invent.kde.org/plasma/kwin/-/merge_requests/3357#note_584723) on how to proceed forward. On one side, I firmly believe that hand-crafted completions are the best quality one can provide, and that worth the effort of some periodic maintenance to keep them in sync with apps. On the other side, the rest of the team is convinced that either parsing `--help` output is a good idea, or investing time in a sky-blue research will lead to some silver-bullet discovery that would allow generating both app's CLI logic code (C++, no less!) and completions for any shells in a generic way. Seems like diplomacy has failed, which leaves me with the only reasonable option: submit my work elsewhere, so it won't rot [in my personal dotfiles](https://github.com/ratijas/dotfiles/blob/1316822abb063f79660ba4a60f02caa0fa2630fe/tag-zsh/config/zsh/functions/Qt/_kscreen-doctor) repo.

Question: will this project accept my work (under GPL-2.0-or-later terms), even though knowing that it was rejected upstream and has little chances of ever moving there?

-- ivan (@ratijas)

**Comment by Freed-Wu** [comment]
I also have [many hand-crafted completions](https://github.com/Freed-Wu/completions) but I think hand-crafted is not a good idea. Many languages has their own library to generate zsh completion script automatically: 

- [rust/clap](https://docs.rs/clap_complete/latest/clap_complete)
- [go/cobra](https://jmtirado.net/shell-completion-with-cobra-and-go/)
- [python/shtab](https://docs.iterative.ai/shtab)
- [nodejs/omelette](https://github.com/f/omelette)

I think generate shell completion script is better. However, a not good solution is better than nothing :smile:

**Comment by msva** [comment]
> even though knowing that it was rejected upstream and has little chances of ever moving there

Actually, this will improve chances for it to be accepted here, as "not being in upstream" is exactly the requirement for completions to be accepted here

**Comment by Freed-Wu** [comment]
> "not being in upstream" is exactly the requirement for completions to be accepted here

True. And if the upstream accepts them later, they will be deleted here like [this](https://github.com/zsh-users/zsh-completions/pull/895).

**Comment by ratijas** [comment]
Hi, it's nice to know you, and thanks for your feedback!

Sounds like it's pretty much settled then.  I'll double-check on my side to make sure everyone is OK with this, before making final decision.
